x/staking/src/types/domain.rs

```diff
@@ -0,0 +1,329 @@
+/// bond shares is based on the amount of coins delegated divided by the current
+/// exchange rate. Voting power can be calculated as total bonded shares
+/// multiplied by exchange rate.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Validator {
+    /// operator_address defines the address of the validator's operator; bech encoded in JSON.
+    pub operator_address: ValAddress,
+    /// delegator_shares defines total shares issued to a validator's delegators.
+    pub delegator_shares: Decimal256,
+    /// description defines the description terms for the validator.
+    pub description: Description,
+    /// consensus_pubkey is the consensus public key of the validator, as a Protobuf Any.
+    pub consensus_pubkey: PublicKey,
+    /// jailed defined whether the validator has been jailed from bonded status or not.
+    pub jailed: bool,
+    /// tokens define the delegated tokens (incl. self-delegation).
+    pub tokens: Uint256,
+    /// unbonding_height defines, if unbonding, the height at which this validator has begun unbonding.
+    pub unbonding_height: i64,
+    /// unbonding_time defines, if unbonding, the min time for the validator to complete unbonding.
+    pub unbonding_time: Timestamp,
+    /// commission defines the commission parameters.
+    pub commission: Commission,
+    pub min_self_delegation: Uint256,
+    pub status: BondStatus,
+}
+
+impl Validator {
+    pub fn new_with_defaults(
+        operator_address: ValAddress,
+        consensus_pubkey: PublicKey,
+        description: Description,
+    ) -> Validator {
+        Validator {
+            operator_address,
+            delegator_shares: Decimal256::zero(),
+            description,
+            consensus_pubkey,
+            jailed: false,
+            tokens: Uint256::zero(),
+            unbonding_height: 0,
+            unbonding_time: Timestamp {
+                seconds: 0,
+                nanos: 0,
+            },
+            commission: Commission {
+                commission_rates: CommissionRates {
+                    rate: Decimal256::zero(),
+                    max_rate: Decimal256::zero(),
+                    max_change_rate: Decimal256::zero(),
+                },
+                update_time: Timestamp {
+                    seconds: 0,
+                    nanos: 0,
+                },
+            },
+            min_self_delegation: Uint256::one(),
+            status: BondStatus::Unbonded,
+        }
+    }
+
+    pub fn abci_validator_update(&self, power: i64) -> ValidatorUpdate {
+        ValidatorUpdate {
+            pub_key: self.consensus_pubkey.clone(),
+            power: self.consensus_power(power),
+        }
+    }
+    pub fn abci_validator_update_zero(&self) -> ValidatorUpdate {
+        self.abci_validator_update(0)
+    }
+
+    pub fn set_initial_commission(&mut self, commission: Commission) -> Result<(), AppError> {
+        commission.validate()?;
```

Comment:
Can we make it so that we're not able to construct a `Comission` which isn't valid?

Reply:
Yes, we can. We can make private fields for the commission and provide a constructor.

Reply:
Ok, sounds good, let's do that

Reply:
updated realization